### PR TITLE
Copied Ports class from the Takes project. Issue #615

### PR DESCRIPTION
**Pull request by antonini**
Copied Ports class from the Takes project as requested on issue #615.

### CODE REVIEW

**Comment by karato** [comment]
@antonini I will find a reviewer for your pull requests shortly, thanks for contribution!

**Comment by karato** [comment]
@caarlos0 review it

**Comment by caarlos0** [comment]
@antonini no tests?

**Comment by antonini** [comment]
@caarlos0 No. It should be used inside tests like on Takes project.

**Comment by caarlos0** [comment]
@antonini OK then.

**Comment by caarlos0** [comment]
@rultor good to merge

File: netbout-web/src/main/java/com/netbout/misc/Ports.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.concurrent.ConcurrentSkipListSet;
+
+/**
+ * Random TCP ports.
+ *
+ * <p>The class is immutable and thread-safe.</p>
+ *
+ * @author Yegor Bugayenko (yegor@teamed.io)
+ * @version $Id$
+ * @since 0.14
```
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/misc/Ports.java]: @antonini should be 2.14.17

File: netbout-web/src/main/java/com/netbout/misc/package-info.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+/**
+ * Misc.
+ *
+ * @author Endrigo Antonini (teamed@endrigo.com.br)
+ * @version $Id$
+ * @since 2.15
```
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/misc/package-info.java]: @antonini same here

File: netbout-web/src/main/java/com/netbout/misc/Ports.java (hunk 1 above)
Comment (antonini) [inline netbout-web/src/main/java/com/netbout/misc/Ports.java]: @dmzaytsev Ok. Going to change that. Where do I find this value? There is no milestone with that value.

**Comment by dmzaytsev** [comment]
@antonini I am with @caarlos0  on this, please add the unit test

File: netbout-web/src/main/java/com/netbout/misc/Ports.java (hunk 1 above)
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/misc/Ports.java]: @antonini latest release here https://github.com/yegor256/netbout/releases
just increment it

Comment (antonini) [inline netbout-web/src/main/java/com/netbout/misc/Ports.java]: @dmzaytsev Ok. Changed to that value. But I found on the source code 2.15 value too.
https://github.com/yegor256/netbout/blob/f464a3becb615db127bd0ed8a32d465d4d82be9b/netbout-web/src/main/java/com/netbout/rest/FkRegistered.java#L43
https://github.com/yegor256/netbout/blob/f464a3becb615db127bd0ed8a32d465d4d82be9b/netbout-web/src/test/java/com/netbout/rest/bout/TkIndexTest.java#L45
Should I report it?

Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/misc/Ports.java]: @antonini why not, but keep in mind this http://www.teamed.io/policy.html#13
this is a cosmetic bug.

**Comment by antonini** [comment]
@dmzaytsev and @caarlos0 Added unit test as required and also fixed since version.

**Comment by dmzaytsev** [comment]
@caarlos0 please review the changed code

File: netbout-web/src/test/java/com/netbout/misc/PortsTest.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        MatcherAssert.assertThat(portb, Matchers.not(portc));
+    }
+
+    /**
+     * Check if generates a random number and doesn't duplicate port with
+     * same Ports instance.
+     * @throws Exception If there is some problem inside
+     */
+    @Test
+    public void checkDifferentNumbersWithSamePorts() throws Exception {
```
Comment (caarlos0) [inline netbout-web/src/test/java/com/netbout/misc/PortsTest.java]: @antonini `allocatesDifferentNumbersWithSamePorts` (check other test classes for examples)

File: netbout-web/src/test/java/com/netbout/misc/PortsTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ */
+public final class PortsTest {
+
+    /**
+     * Check if generates a random number and doesn't duplicate port with
+     * different Ports instance.
+     * @throws Exception If there is some problem inside
+     */
+    @Test
+    public void checkDifferentNumbersWithDifferentPorts() throws Exception {
```
Comment (caarlos0) [inline netbout-web/src/test/java/com/netbout/misc/PortsTest.java]: @antonini `allocatesDifferentNumbersWithDifferentPorts` (check other test classes for examples)

File: netbout-web/src/test/java/com/netbout/misc/PortsTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * Test case for {@link Ports}.
+ *
+ * @author Endrigo Antonini (teamed@endrigo.com.br)
+ * @version $Id$
+ * @since 2.14.17
+ */
+public final class PortsTest {
+
+    /**
+     * Check if generates a random number and doesn't duplicate port with
```
Comment (caarlos0) [inline netbout-web/src/test/java/com/netbout/misc/PortsTest.java]: @antonini `Ports can genereate...` (check other test classes for examples)

File: netbout-web/src/test/java/com/netbout/misc/PortsTest.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        final int porta = new Ports().allocate();
+        final int portb = new Ports().allocate();
+        final int portc = new Ports().allocate();
+        MatcherAssert.assertThat(porta, Matchers.not(portb));
+        MatcherAssert.assertThat(porta, Matchers.not(portc));
+        MatcherAssert.assertThat(portb, Matchers.not(portc));
+    }
+
+    /**
+     * Check if generates a random number and doesn't duplicate port with
```
Comment (caarlos0) [inline netbout-web/src/test/java/com/netbout/misc/PortsTest.java]: @antonini `Ports can genereate...` (check other test classes for examples)

**Comment by caarlos0** [comment]
@antonini a couple of comments above..

**Comment by antonini** [comment]
@caarlos0 Fixed javadoc and also method name. Can you review?

**Comment by caarlos0** [comment]
@antonini thanks!

**Comment by caarlos0** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@rultor merge please

@antonini @caarlos0 thanks!

**Comment by karato** [comment]
@ypshenychka please, let us know what do you think about this ticket, according to our [QA rules](http://www.teamed.io/qa.html)

**Comment by karato** [comment]
@rultor deploy now pls

**Comment by ypshenychka** [comment]
@dmzaytsev Please advise here. To my mind the issues found by code reviwer here are mostly cosmetic. Am I right?

**Comment by dmzaytsev** [comment]
@ypshenychka the problems above is definitely not a design problem except the lack the unit tests.
To my mind the test method name is a cosmetic issue. But according [this paper](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names) we can't accept PR with an invalid test method name.

**Comment by ypshenychka** [comment]
@dmzaytsev Thank you for support.

**Comment by ypshenychka** [comment]
@caarlos0 
According to our [QA Rules](http://www.teamed.io/qa.html):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one major issue was revealed during code review. Try to find at least three design problems while future reviews.

Please confirm that everything is clear and you have no questions regarding this approach.

**Comment by caarlos0** [comment]
@ypshenychka yes, I'll try

**Comment by ypshenychka** [comment]
@caarlos0 Thanks.

**Comment by ypshenychka** [comment]
@karato Quality is acceptable.

**Comment by karato** [comment]
> @karato Quality is acceptable.

@ypshenychka thanks for the review, we'll try better next time

**Comment by karato** [comment]
@caarlos0 10 mins was added to the account of @ypshenychka (for QA review), in transaction 58522749

Many thanks! **22 mins** were added to your account in Transaction ID `AP-7HT73054F1809653L` (task took 49 hours and 13 mins)

review comments (c=7) added as a bonus

+22 added to your rating, current score is: [+2042](http://www.netbout.com/b/35095?open=rating)

**Pull request merged**